_posts/2018-08-22-optimize-performance-in-docker.md

```diff
@@ -0,0 +1,219 @@
+No `ccache` single-threaded:
+
+```
+docker run --rm -it -v $PWD:/home/xen -w /home/xen 3mdeb/xen-docker make debball| ts -s '[%.T]'
+(...)
+[00:13:10.006206] dpkg-deb: building package 'xen-upstream' in 'xen-upstream-4.8.4.deb'.
+```
+
+Cold cache:
+
+```
+```
+
+Hot cache:
+
+```
+
+```
+
+# apt-cacher-ng
+
+There 2 use case for `apt-cacher-ng` in our workflows. One is Docker build
+time, which can be time-consuming since all packages and its dependencies are
+installed in the base image. Second is runtime, when you need some package that
+may have extensive dependencies e.g. `xen-systema-amd64`.
+
+First, let's setup `apt-cacher-ng`. Some guide may be found in [Docker documentation](https://docs.docker.com/engine/examples/apt-cacher-ng/), but we will modify it a little bit.
+
+Ideally, we would like to use `docker compose` to set up `apt-cacher-ng`
+container whenever it is not set, or have dedicated VM which serves this
+purpose. In this post, we consider local cache. Dockerfile may look like this:
+
+```
+FROM        ubuntu
+
+RUN     apt-get update && apt-get install -y apt-cacher-ng
+
+EXPOSE      3142
+CMD     chmod 777 /var/cache/apt-cacher-ng && /etc/init.d/apt-cacher-ng start && tail -f /var/log/apt-cacher-ng/*
+```
+
+Build and run:
+
+```
+docker build -t apt-cacher .
+docker run -d -p 3142:3142 -v $PWD/apt_cache:/var/cache/apt-cacher-ng --name cacher-container apt-cacher
+docker logs -f cacher-container
+```
+
+The output should look like this:
+
+```
+* Starting apt-cacher-ng apt-cacher-ng
+WARNING: No configuration was read from file:sfnet_mirrors
+   ...done.
+   ==> /var/log/apt-cacher-ng/apt-cacher.err <==
+
+   ==> /var/log/apt-cacher-ng/apt-cacher.log <==
+```
+
+We should also see that cacher listens on port `3142`:
+
+```
+[16:40:01] pietrushnic:~ $ netstat -an |grep 3142
+tcp6       0      0 :::3142                 :::*                    LISTEN
+```
+
+Now we can run docker building with appropriate parameters:
+
+```
+docker build --build-arg HTTP_PROXY=http://192.168.4.112:3142/ -t 3mdeb/xen-docker .| ts -s '[%.T]'
+```
+
+## performance measures
+
+`xen-docker` container build without `apt-cacher-ng`. To measure what is going
+on during container build we are using `ts` from `moreutils` package.
+
+Without cacher:
+
+```
+docker build -t 3mdeb/xen-docker .| ts -s '[%.S]'
+(...)
+[00:07:13.723282] Successfully tagged 3mdeb/xen-docker:latest
+```
+
+With cold cacher:
+
+```
+[00:06:55.051968] Successfully tagged 3mdeb/xen-docker:latest
+```
+
+With hot cacher:
+
+```
+docker build --build-arg http_proxy=http://<CACHER_IP>:3142/ -t 3mdeb/xen-docker .| ts -s '[%.T]'
+(...)
+[00:05:50.237480] Successfully tagged 3mdeb/xen-docker:latest
+```
+
+Assuming that the network conditions did not change between runs to extent of
+30s delay we can conclude:
+
+* using cacher even with cold cache is better than nothing, it gives the
+  speedup of about 5%.
+* using filled cache can give ~20% over container build time, if significant
```

Comment:
@pietrushnic The `filled` cache concept was not mentioned before

Reply:
@macpijan I assumed this is obvious :) Of course I meant `hot` instead of `filled`.